python/ray/serve/_private/metrics_utils.py

```diff
@@ -165,52 +178,204 @@ def _get_datapoints(
+        Args:
+            keys: Iterable of keys to aggregate across.
+        Returns:
+            A tuple of (float, int) where the first element is the mean across
+            all values found at `keys`, and the second is the number of valid
+            keys used to compute the mean.
+            Returns (None, 0) if no valid keys have data.
+        """
+        return self._aggregate_reduce(keys, statistics.mean)
+
+
+def _bucket_latest_by_window(
+    series: List[TimeStampedValue],
+    start: float,
+    window_s: float,
+) -> Dict[int, float]:
+    """
+    Map each window index -> latest value seen in that window.
+    Assumes series is sorted by timestamp ascending.
+    """
+    buckets: Dict[int, float] = {}
+    for p in series:
+        w = int((p.timestamp - start) // window_s)
+        buckets[w] = p.value  # overwrite keeps the latest within the window
+    return buckets
+
+
+def _merge_two_timeseries(
+    t1: List[TimeStampedValue], t2: List[TimeStampedValue], window_s: float
+) -> List[TimeStampedValue]:
+    """
+    Merge two ascending time series by summing values within a specified time window.
+    If multiple values fall within the same window in a series, the latest value is used.
+    The output contains one point per window that had at least one value, timestamped
+    at the window center.
+    """
+    if not t1 and not t2:
+        return []
+
+    # Align windows so each output timestamp sits at the center of its window.
+    earliest = min(x[0].timestamp for x in (t1, t2) if x)
+    start = earliest - (window_s / 2.0)
+
+    b1 = _bucket_latest_by_window(t1, start, window_s)
+    b2 = _bucket_latest_by_window(t2, start, window_s)
+
+    windows = sorted(set(b1.keys()) | set(b2.keys()))
+
+    merged: List[TimeStampedValue] = []
+    for w in windows:
+        v = b1.get(w, 0.0) + b2.get(w, 0.0)
+        ts_center = start + w * window_s + (window_s / 2.0)
+        merged.append(TimeStampedValue(timestamp=ts_center, value=v))
+    return merged
+
+
+def merge_timeseries_dicts(
+    *timeseries_dicts: DefaultDict[Hashable, List[TimeStampedValue]],
+    window_s: float,
+) -> DefaultDict[Hashable, List[TimeStampedValue]]:
+    """
+    Merge multiple time-series dictionaries, typically contained within
+    InMemoryMetricsStore().data. For the same key across stores, time series
+    are merged with a windowed sum, where each series keeps only its latest
+    value per window before summing.
+    """
+    if len(timeseries_dicts) == 1:
+        return timeseries_dicts[0]
+
+    merged: DefaultDict[Hashable, List[TimeStampedValue]] = defaultdict(list)
+    for timeseries_dict in timeseries_dicts:
+        for key, ts in timeseries_dict.items():
+            if key in merged:
+                merged[key] = _merge_two_timeseries(merged[key], ts, window_s)
+            else:
+                # Keep raw data for unique keys
+                merged[key] = ts.copy()
```

Comment:
@vaishdho1 raised a good point in https://github.com/ray-project/ray/pull/55568#discussion_r2299242202. How about we do something like this:
```suggestion
            merged[key] = _merge_two_timeseries(merged.get(key, []), ts, window_s)
```
and have _merge_two_timeseries gracefully handle empty lists?